src/Plugin/Field/FieldWidget/OgComplex.php

```diff
@@ -26,13 +31,310 @@
 class OgComplex extends EntityReferenceAutocompleteWidget {
 
   /**
-   * The OG complex widget have a special logic on order to return the groups
-   * that user can reference to.
+   * {@inheritdoc}
    */
   public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
     $parent = parent::formElement($items, $delta, $element, $form, $form_state);
+    // todo: fix the definition in th UI level.
     $parent['target_id']['#selection_handler'] = 'default:og';
+    $parent['target_id']['#selection_settings']['field_mode'] = 'default';
+
     return $parent;
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function form(FieldItemListInterface $items, array &$form, FormStateInterface $form_state, $get_delta = NULL) {
+    $parent_form = parent::form($items, $form, $form_state, $get_delta);
+
+    $parent_form['other_groups'] = [];
+
+    // Adding the other groups widget.
+    if ($this->isGroupAdmin()) {
+      $parent_form['other_groups'] = $this->otherGroupsWidget($items, $form_state);
+    }
+
+    return $parent_form;
+  }
+
+  /**
+   * Special handling to create form elements for multiple values.
+   *
+   * Handles generic features for multiple fields:
+   * - number of widgets
+   * - AHAH-'add more' button
+   * - table display and drag-n-drop value reordering
+   */
+  protected function formMultipleElements(FieldItemListInterface $items, array &$form, FormStateInterface $form_state) {
+    $field_name = $this->fieldDefinition->getName();
+    $cardinality = $this->fieldDefinition->getFieldStorageDefinition()->getCardinality();
+    $parents = $form['#parents'];
+
+    $target_type = $this->fieldDefinition->getTargetEntityTypeId();
+    $user_groups = Og::getEntityGroups('user', \Drupal::currentUser()->id());
+    $user_groups_target_type = isset($user_groups[$target_type]) ? $user_groups[$target_type] : [];
+    $user_group_ids = array_map(function($group) {
+      return $group->id();
+    }, $user_groups_target_type);
+
+    // Determine the number of widgets to display.
+    switch ($cardinality) {
+      case FieldStorageDefinitionInterface::CARDINALITY_UNLIMITED:
+        $field_state = static::getWidgetState($parents, $field_name, $form_state);
+        $max = $field_state['items_count'];
+        $is_multiple = TRUE;
+        break;
+
+      default:
+        $max = $cardinality - 1;
+        $is_multiple = ($cardinality > 1);
+        break;
+    }
+
+    $title = $this->fieldDefinition->getLabel();
+    $description = FieldFilteredMarkup::create(\Drupal::token()->replace($this->fieldDefinition->getDescription()));
+
+    $elements = array();
+
+    for ($delta = 0; $delta <= $max; $delta++) {
+      // Add a new empty item if it doesn't exist yet at this delta.
+      if (!isset($items[$delta])) {
+        $items->appendItem();
+      }
+      elseif (!in_array($items[$delta]->get('target_id')->getValue(), $user_group_ids)) {
```

Comment:
@amitaibu for now I think we should go with this, we have to copy the formMultipleElements method so that we can add this change to skip non user groups. Otherwise I think we're in a world of pain here...